### DISCUSSION TITLE
[warning] Mapping is not an object?

### DISCUSSION THREAD

**Discussion by Sibuli** [Alexays/Waybar#2816] [Q&A]
Hi!

I didn't find anyone mention this or any fixes for it. Just installed linux a few days ago, so I'm really new to all of this, might also be that I did something wrong in the installation process, but the problem occurs when I start waybar. I tried to follow arch wiki with the installation and other programs run fine so I'm guessing it's in waybar. When I start waybar, it says:
"[warning] Mapping is not an object"
No other warnings or errors atm, had a few but got them working/clear. No luck with this one though. Note that waybar opens and works, but after launch, terminal is "frozen" or stuck processing or whatever. I can exit with Ctrl+c but it shuts down waybar. Do you have any idea what this error is, or if there's a problem in my config?

(EDIT: Theres also this notification when opening waybar, maybe important maybe not)
 [info] Unable to receive desktop appearance: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface “org.freedesktop.portal.Settings” on object at path /org/freedesktop/portal/desktop

OS: Arch Linux x86_64 
Host: 82MS Yoga Slim 7 Pro 14ACH5 
Kernel: 6.6.10-arch1-1 
Shell: bash 5.2.21 
Hyprland 
Terminal: kitty 
CPU: AMD Ryzen 5 5600H with Radeon Graphics (12) @ 4.280GHz 
GPU: AMD ATI Radeon Vega Series / Radeon Vega Mobile Series 

my config file:

{
     "layer": "top",
     "position": "top", // Waybar position (top|bottom|left|right)
    "height": 25, // Waybar height (to be removed for auto height)
    // "width": 1280, // Waybar width
    "spacing": 4, // Gaps between modules (4px)

    "modules-left": [ "hyprland/workspaces" ],
    "modules-center": [ "clock" ],
    "modules-right": [ 
    "pulseaudio",
    "network", 
    "cpu", 
    "memory", 
    "temperature", 
    //"backlight", 
    "battery" ],

     "hyprland/workspaces": {
         "disable-scroll": true,
         "all-outputs": true,
         "warp-on-scroll": false,
         "format": "{name}: {icon}",
         "format-icons": {
    //         "1": "",
    //         "2": "",
    //         "3": "",
    //         "4": "",
    //         "5": "",
    //         "urgent": "",
    //         "focused": "",
             "default": ""
        }
    },
    //"hyprland/scratchpad": {
      //  "format": "{icon} {count}",
      //  "show-empty": false,
      // "format-icons": ["", ""],
      //  "tooltip": true,
      //  "tooltip-format": "{app}: {title}"
      // },
  //  "mpd": {
  //      "format": "{stateIcon} {consumeIcon}{randomIcon}{repeatIcon}{singleIcon}{artist} - {album} - {title} ({elapsedTime:%M:%S}/{totalTime:%M:%S}) ⸨{songPosition}|{queueLength}⸩ {volume}% ",
  //      "format-disconnected": "Disconnected ",
  //      "format-stopped": "{consumeIcon}{randomIcon}{repeatIcon}{singleIcon}Stopped ",
  //      "unknown-tag": "N/A",
  //      "interval": 2,
  //      "consume-icons": {
  //          "on": " "
  //      },
  //      "random-icons": {
  //          "off": "<span color=\"#f53c3c\"></span> ",
  //          "on": " "
  //      },
  //      "repeat-icons": {
  //          "on": " "
  //      },
  //      "single-icons": {
  //          "on": "1 "
  //      },
  //      "state-icons": {
  //          "paused": "",
  //          "playing": ""
  //      },
  //      "tooltip-format": "MPD (connected)",
  //      "tooltip-format-disconnected": "MPD (disconnected)"
  //  },
  //  "idle_inhibitor": {
  //      "format": "{icon}",
  //      "format-icons": {
  //          "activated": "",
  //          "deactivated": ""
  //      }
  //  },
  //  "tray": {
  //      // "icon-size": 21,
  //      "spacing": 10
  //  },
    "clock": {
        // "timezone": "America/New_York",
        "tooltip-format": "<big>{:%Y %B}</big>\n<tt><small>{calendar}</small></tt>",
        "format-alt": "{:%Y-%m-%d}"
    },
    "cpu": {
        "format": "{usage}% ",
        "tooltip": false
    },
    "memory": {
        "format": "{}% "
    },
    "temperature": {
         "thermal-zone": 2,
         "hwmon-path": "/sys/class/hwmon/hwmon2/temp1_input",
        "critical-threshold": 80,
         "format-critical": "{temperatureC}°C {icon}",
        "format": "{temperatureC}°C {icon}",
        "format-icons": ["", "", ""]
    },
    "backlight": {
         "device": "acpi_video1",
        "format": "{percent}% {icon}",
        "format-icons": ["", "", "", "", "", "", "", "", ""]
    },
    "battery": {
        "states": {
             "good": 95,
            "warning": 30,
            "critical": 15
        },
        "format": "{capacity}% {icon}",
        "format-charging": "{capacity}% ",
        "format-plugged": "{capacity}% ",
        "format-alt": "{time} {icon}",
        // "format-good": "", // An empty format will hide the module
        // "format-full": "",
        "format-icons": ["", "", "", "", ""]
    },
    "network": {
        // "interface": "wlp2*", // (Optional) To force the use of this interface
        "format-wifi": "{essid} ({signalStrength}%) ",
        "format-ethernet": "{ipaddr}/{cidr} ",
        "tooltip-format": "{ifname} via {gwaddr} ",
        "format-linked": "{ifname} (No IP) ",
        "format-disconnected": "Disconnected ⚠",
        "format-alt": "{ifname}: {ipaddr}/{cidr}"
    },
    "pulseaudio": {
        // "scroll-step": 1, // %, can be a float
        "format": "{volume}% {icon} {format_source}",
        "format-bluetooth": "{volume}% {icon} {format_source}",
        "format-bluetooth-muted": " {icon} {format_source}",
        "format-muted": " {format_source}",
        "format-source": "{volume}% ",
        "format-source-muted": "",
        "format-icons": {
            "headphone": "",
            "hands-free": "",
            "headset": "",
            "phone": "",
            "portable": "",
            "car": "",
            "default": ["", "", ""]
        },
        "on-click": "pavucontrol"
    },
    "custom/media": {
        "format": "{icon} {}",
        "return-type": "json",
        "max-length": 40,
        "format-icons": {
            "spotify": "",
            "default": "🎜"
        },
        "escape": true,
        "exec": "$HOME/.config/waybar/mediaplayer.py 2> /dev/null" // Script in resources folder
        // "exec": "$HOME/.config/waybar/mediaplayer.py --player spotify 2> /dev/null" // Filter player based on name
    }
}

**Comment by ignamartinoli** [comment]
I receive the same message with 

```json
{
	"layer": "top",
	"margin-top": 6,
	"margin-right": 10,
	"margin-bottom": 0,
	"margin-left": 10,
	"spacing": 5,
	"modules-left": [ "hyprland/workspaces", "hyprland/submap" ],
	"modules-center": [ "clock" ],
	"modules-right": [ "custom/spotify", "wireplumber", "backlight", "network", "battery" ],
	"hyrpland/workspaces": {
		"all-outputs": false,
		"disable-scroll": true,
		"on-click": "activate"
	},
	"hyprland/submap": {
		"format": "R",
		"tooltip": false
	},
	"clock": {
		"format": "{:%H:%M }",
		"format-alt": "{:%a, %b %dth %Y }",
		"timezone": "America/Argentina/Cordoba",
		"tooltip-format": "<big>{:%Y %B \t   week: %V }</big>\n<tt>{calendar}</tt>"
	},
	"custom/spotify": {
		"exec": "$HOME/.config/waybar/music.sh 2> /dev/null",
		"exec-if": "pgrep '^spotify$'",
		"format": "{}",
		"on-click": "playerctl -p 'spotify' play-pause",
		"on-click-right": "playerctl -p 'spotify' next",
		"tooltip": false
	},
	"wireplumber": {
		"format": "{volume}% ",
		"format-muted": "Muted ",
		"on-click": "wpctl set-mute @DEFAULT_AUDIO_SINK@ toggle",
		"states": [ "", "" ]
	},
	"backlight": {
		"format": "{percent}% {icon}",
		"format-icons": [ "", "", "", "" ],
		"on-click": "light -Ur 16",
		"on-click-right": "light -Ar 16",
		"on-scroll-down": "light -Ur 1",
		"on-scroll-up": "light -Ar 1",
		"tooltip": false
	},
	"network": {
		"format-alt": "{essid} ({signalStrength}%) ",
		"format-disconnected": "Disconnected ",
		"format-wifi": "{essid} ",
		"interface": "wlan*",
		"tooltip": false
	},
	"battery": {
		"format": "{capacity}% {icon}",
		"format-charging": "{capacity}% ",
		"format-icons": [ "", "", "", "", "" ],
		"states": {
			"critical": 10
		}
	}
}
```

**Comment by alekgr** [comment]
Yes I am get this message along with

```
 [2024-03-06 07:33:10.248] [info] Using configuration file /home/alek/.config/waybar/config
[2024-03-06 07:33:10.249] [info] Unable to receive desktop appearance: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface “org.freedesktop.portal.Settings” on object at path /org/freedesktop/portal/desktop
[2024-03-06 07:33:10.249] [info] Using CSS file /home/alek/.config/waybar/style.css
[2024-03-06 07:33:10.253] [warning] Mapping is not an object
[2024-03-06 07:33:10.253] [info] Hyprland IPC starting
[2024-03-06 07:33:10.306] [info] Found device AT Translated Set 2 keyboard at '/dev/input/event2'
[2024-03-06 07:33:10.433] [info] Found device Logitech USB Receiver at '/dev/input/event7'
[2024-03-06 07:33:10.599] [info] Found device Keychron Keychron K2 at '/dev/input/event14'
[2024-03-06 07:33:11.149] [warning] Requested height: 23 is less than the minimum height: 32 required by the modules
[2024-03-06 07:33:11.149] [info] Bar configured (width: 3840, height: 32) for output: HDMI-A-1
[2024-03-06 07:33:11.149] [warning] Requested height: 15 is less than the minimum height: 30 re

```


Here is my

**Comment by ikalco** [comment]
To fix it just make sure you have the ["window-rewrite"](https://github.com/Alexays/Waybar/blob/4c46d7d245a6c06644d6a0e8857f7140556202ce/man/waybar-hyprland-workspaces.5.scd?plain=1#L24-L28) property set to an object. inside "hyprland/workspaces".
```
"hyrpland/workspaces": {
	"window-rewrite": {},
	// The rest of your stuff for hyprland/workspaces
},
```
### Explanation
I looked at the source code and found the root of the problem. What's basically happening is that whenever Waybar parses the config for "hyprland/workspaces" and it can't find the "window-rewrite" property in it, the it just defaults it to a string "?". 
https://github.com/Alexays/Waybar/blob/4c46d7d245a6c06644d6a0e8857f7140556202ce/src/modules/hyprland/workspaces.cpp#L130-L142

Which is not an object, as was expected, and therefore causes RegexCollection to print the warning when it is instantiated.
https://github.com/Alexays/Waybar/blob/4c46d7d245a6c06644d6a0e8857f7140556202ce/src/util/regex_collection.cpp#L10-L17

I'm not really familiar with the codebase so I'm not going to try to fix this just wanted to put this here as a pointer if anyone else wanted to try to fix the code.

**Reply by gabyx** [reply]
Solves this also for `sway/workspaces`.

**Reply by ghost** [reply]
oh my god this was driving me crazy. cheers

**Comment by alekgr** [comment]
@ikalco  thank you for looking at the source code. So I am looking at the man page
I only see waybar-wlr-workspaces but I am using hyperland/workspaces and I don't see window-rewrite anywhere

I think they are both same modules but maybe documentation has been  updated

**Comment by ikalco** [comment]
@alekgr yeah I don't see waybar-hyprland-workspaces in my man pages either but it is in the [github](https://github.com/Alexays/Waybar/blob/master/man/waybar-hyprland-workspaces.5.scd) which makes me think it has something to do with Waybar's man pages not being installed fully.
